### PR TITLE
refactor asset bundle code to not depend on the global Cache.flutterRoot

### LINKED ISSUE #142285: `asset.dart` subtly depends on a global via `Cache.flutterRoot`, inhibiting hermetic tests of asset bundle code
`ManifestAssetBundle` accepts dependencies via its constructor to allow for hermetic testing of its behavior:

https://github.com/flutter/flutter/blob/0833929c997c8a9db11c1e0df9a731ab17b77606/packages/flutter_tools/lib/src/asset.dart#L134-L139

However, we refer to the global `Cache.flutterRoot` in a couple places throughout the code:

https://github.com/flutter/flutter/blob/0833929c997c8a9db11c1e0df9a731ab17b77606/packages/flutter_tools/lib/src/asset.dart#L561

https://github.com/flutter/flutter/blob/0833929c997c8a9db11c1e0df9a731ab17b77606/packages/flutter_tools/lib/src/asset.dart#L576-L579

This means that tools test using `testWithoutContext` have to remember to manually set this global. Example:

https://github.com/flutter/flutter/blob/0833929c997c8a9db11c1e0df9a731ab17b77606/packages/flutter_tools/test/general.shard/asset_bundle_test.dart#L382-L389

If nothing worse, this is noisy. Let's consider having `ManifestAssetBundle` accept the flutter root as a constructor parameter.

### PR DESCRIPTION
**Pull request by andrewkolos**
Fixes https://github.com/flutter/flutter/issues/142285.

Part of work on https://github.com/flutter/flutter/pull/141194.

This is a refactor. There should be no changes in tool behavior.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by andrewkolos** [comment]
I'd like to delete `Cache.flutterRoot` entirely, but I'm leaving this out of scope for now. This work is already tangential from my current work—I just want to get `ManifestAssetBundle` to a state where it can be properly unit tested.

File: packages/flutter_tools/lib/src/asset.dart [hunk 1]
```diff
@@ -64,7 +64,13 @@ abstract class AssetBundleFactory {
     required FileSystem fileSystem,
     required Platform platform,
     bool splitDeferredAssets = false,
-  }) => _ManifestAssetBundleFactory(logger: logger, fileSystem: fileSystem, platform: platform, splitDeferredAssets: splitDeferredAssets);
+  }) =>
+      _ManifestAssetBundleFactory(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/asset.dart:68]: nit this should be on the previous line

File: packages/flutter_tools/lib/src/asset.dart [hunk 2]
```diff
@@ -138,7 +144,13 @@ class _ManifestAssetBundleFactory implements AssetBundleFactory {
   final bool _splitDeferredAssets;
 
   @override
-  AssetBundle createBundle() => ManifestAssetBundle(logger: _logger, fileSystem: _fileSystem, platform: _platform, splitDeferredAssets: _splitDeferredAssets);
+  AssetBundle createBundle() => ManifestAssetBundle(
+        logger: _logger,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/asset.dart:148]: nit. this should be indented previous indent + 2: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#indent-multi-line-argument-and-parameter-lists-by-2-characters

**Review by christopherfujino — APPROVED** [review]
nice!

File: packages/flutter_tools/lib/src/asset.dart (hunk 2 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/asset.dart:148]: Yeah, I didn't check the autoformatter's work here

**Pull request merged**